File: api/src/main/kotlin/org/team619/rutor/converter/RowConverter.java
package org.team619.rutor.converter;

import org.jsoup.nodes.Element;
import org.team619.rutor.core.Logger;
import org.team619.rutor.model.Row;

import java.io.Serializable;
import java.util.List;

/**
 * Created by BORIS on 27.08.2016.
 */
public abstract class RowConverter extends DefaultConverter {

    public RowConverter(Logger logger) {
        super(logger);
    }

    protected Row parseRow(List<Element> torrentRow) {
        Row row = null;

        if (torrentRow.size() >= Constants.SIZE_STANDARD) {
            row = new Row();
            int torrentRowIndex = parseRowUrls(torrentRow, row);

            if (torrentRow.size() == Constants.SIZE_STANDARD) {
                row.setSize(parseString(torrentRow.get(torrentRowIndex++).text()));
            } else {
                if (torrentRow.size() >= Constants.SIZE_WITH_COMMENT) {
                    row.setComments(parseInteger(torrentRow.get(torrentRowIndex++).text()));
                    row.setSize(parseString(torrentRow.get(torrentRowIndex++).text()));
                }
            }

            List<Element> downloadInfoDetails = torrentRow.get(torrentRowIndex++).getElementsByTag(Constants.SPAN);
            if (downloadInfoDetails.size() >= Constants.SIZE_HAS_TOPICS) {
                int downloadInfoDetailsIndex = 0;
                row.setSeeds(parseInteger(downloadInfoDetails.get(downloadInfoDetailsIndex++).text()))
                        .setPeers(parseInteger(downloadInfoDetails.get(downloadInfoDetailsIndex++).text()));
            }
        }

        return row;
    }

    protected int parseRowUrls(List<Element> torrentRow, Row row) {
        int torrentRowIndex = 0;
        row.setCreationDate(torrentRow.get(torrentRowIndex++).text());
        List<Element> captionDetails = torrentRow.get(torrentRowIndex++).getElementsByTag(Constants.LINK);
        if (captionDetails.size() >= Constants.SIZE_CAPTION) {
            int captionDetailsIndex = 0;

            row.setDownloadUrl(captionDetails.get(captionDetailsIndex++).attr(Constants.HREF))
                    .setId(parseId(row.getDownloadUrl()))
                    .setMagnetUrl(captionDetails.get(captionDetailsIndex++).attr(Constants.HREF))
                    .setDetailUrl(captionDetails.get(captionDetailsIndex).attr(Constants.HREF))
                    .setFileName(parseTorrentName(captionDetails.get(captionDetailsIndex).attr(Constants.HREF)))
                    .setCaption(parseCaption(captionDetails.get(captionDetailsIndex++).text()));
        }

        return torrentRowIndex;
    }
    
    /* =================== */

    public interface Constants extends Serializable {

        int SIZE_STANDARD = 4;
        int SIZE_WITH_COMMENT = 5;
        int SIZE_HAS_TOPICS = 2;
        int SIZE_CAPTION = 3;

        String LINK = "a";
        String HREF = "href";
        String SPAN = "span";

    }

}

File: api/src/main/kotlin/org/team619/rutor/converter/RowConverter.kt
package org.team619.rutor.converter

import org.jsoup.nodes.Element
import org.team619.rutor.core.Logger
import org.team619.rutor.model.Row
import org.team619.rutor.util.Util

/**
 * Created by BORIS on 27.08.2016.
 */
abstract class RowConverter(logger: Logger) : DefaultConverter(logger) {

    protected fun parseRow(torrentRow: List<Element>): Row? {
        var row: Row? = null

        if (torrentRow.size >= Constants.SIZE_STANDARD) {
            row = Row()
            var torrentRowIndex = parseRowUrls(torrentRow, row)

            if (torrentRow.size == Constants.SIZE_STANDARD) {
                row.size = Util.parseString(torrentRow[torrentRowIndex++].text())
            } else {
                if (torrentRow.size >= Constants.SIZE_WITH_COMMENT) {
                    row.comments = parseInteger(torrentRow[torrentRowIndex++].text())
                    row.size = Util.parseString(torrentRow[torrentRowIndex++].text())
                }
            }

            val downloadInfoDetails = torrentRow[torrentRowIndex++].getElementsByTag(Constants.SPAN)
            if (downloadInfoDetails.size >= Constants.SIZE_HAS_TOPICS) {
                var downloadInfoDetailsIndex = 0
                row.seeds = parseInteger(downloadInfoDetails[downloadInfoDetailsIndex++].text())
                row.peers = parseInteger(downloadInfoDetails[downloadInfoDetailsIndex++].text())
            }
        }

        return row
    }

    protected fun parseRowUrls(torrentRow: List<Element>, row: Row): Int {
        var torrentRowIndex = 0
        row.creationDate = torrentRow[torrentRowIndex++].text()
        val captionDetails = torrentRow[torrentRowIndex++].getElementsByTag(Constants.LINK)
        if (captionDetails.size >= Constants.SIZE_CAPTION) {
            var captionDetailsIndex = 0

            row.downloadUrl = captionDetails[captionDetailsIndex++].attr(Constants.HREF)
            row.id = Util.parseId(row.downloadUrl)
            row.magnetUrl = captionDetails[captionDetailsIndex++].attr(Constants.HREF)
            row.detailUrl = captionDetails[captionDetailsIndex].attr(Constants.HREF)
            row.fileName = parseTorrentName(captionDetails[captionDetailsIndex].attr(Constants.HREF))
            row.caption = parseCaption(captionDetails[captionDetailsIndex++].text())
        }

        return torrentRowIndex
    }

    interface Constants {

        companion object {

            const val SIZE_STANDARD = 4
            const val SIZE_WITH_COMMENT = 5
            const val SIZE_HAS_TOPICS = 2
            const val SIZE_CAPTION = 3

            const val LINK = "a"
            const val HREF = "href"
            const val SPAN = "span"
        }

    }

}
